impl/src/main/java/org/jboss/weld/bootstrap/WeldStartup.java

```diff
@@ -574,6 +576,7 @@ public void endInitialization() {
             for (BeanDeploymentModule module : modules) {
                 if (!module.isWebModule()) {
                     module.fireEvent(Object.class, ContextEvent.APPLICATION_INITIALIZED, Initialized.Literal.APPLICATION);
+                    module.fireEvent(Startup.class, new Startup(), Any.Literal.INSTANCE);
```

Comment:
Hm, would it make sense to fire the `Startup` event for all modules here? You don't need to fire the event from within  `HttpContextLifecycle` as you don't need to access the `ServletContext`... Actually, it might make sense to create an SPI so that an integrator can decide when to fire this event because otherwise I see no benefit compared to `@Initialized(ApplicationScoped.class)`. WDYT?

Reply:
> Hm, would it make sense to fire the Startup event for all modules here?

It should be fired for all modules - the `HttpContextLifecycle` contains the other part. Or did you mean something else?

> Actually, it might make sense to create an SPI so that an integrator can decide when to fire this event because otherwise I see no benefit compared to @Initialized(ApplicationScoped.class). WDYT?

I tend to disagree. For a runtime implementation such as Weld, it is very much identical to initialization of app scope and that's IMO fine.

These events were added to CDI for two reasons from what I can recall:
* For any build time implementation to allow them to differentiate initialization of app scope from state of actual container readiness as the two might differ. We already saw this in Quarkus.
* To create separate standard events that stand for the init and shutdown of container and don't require preliminary knowledge of other components to make use of them. I.e. you can now observe an event that directly represent container ready state as opposed to observing a context init event where you need to know that it happens as a first thing after container startup

Reply:
> It should be fired for all modules - the HttpContextLifecycle contains the other part. Or did you mean something else?

I thought that you don't need to fire this even from the `HttpContextLifecycle` but instead fire it from this loop outside the `if (!module.isWebModule())` block.

Reply:
Oh I see; sure, can do that.

Reply:
+1 for an SPI. The events need to be fired at slightly different times to the actual init and shutdown of the CDI container. Really these are "post-Startup" and "pre-Shutdown" events.

Reply:
@mkouba @tevans78 how do you imagine this SPI should look like?

Since it is a plain CDI event and integrators have the complete knowledge on when to fire it, do we really need an SPI to achieve that?

Reply:
Technically we tried to formulate the spec of the `Startup` and `Shutdown` events in a way that it allows firing them at the same time as `@Initialized|BeforeDestroyed(ApplicationScoped.class) Object` for the purely runtime implementations like Weld. At the same time, I agree it would be great for integrators to be able to indicate to Weld that it should not fire these events, because the integrator will do it itself (ideally at a better point in time :-) ).

Reply:
@manovotn I was thinking so something really simple like `WeldBootstrap.fireStartupEvent() / fireShutdownEvent()`. However, If the web modules are not processed separately via the HttpContextLifecycle, and there is a way for us to prevent Weld from firing the events, then perhaps you are correct that we could fire these events independently. I will experiment with that.

Regardless of how it is done, getting the timing right could be difficult/unclear. Just because the CDI container has been started, should that guarantee that all other JEE services are ready and started... but no requests will be processed yet? On Shutdown, should everything else still be available... but no more requests be processed?

Reply:
> I was thinking so something really simple like `WeldBootstrap.fireStartupEvent()` / `fireShutdownEvent()`...

One thing I don't know from the top of my head is if you can fire that event and be able to observe it globally or whether you need to fire it in some per-module manner. I'd need to check tomorrow. Do let me know if you experiment with firing the event yourself please.

> Regardless of how it is done, getting the timing right could be difficult/unclear. Just because the CDI container has been started, should that guarantee that all other JEE services are ready and started... but no requests will be processed yet? On Shutdown, should everything else still be available... but no more requests be processed?

This is hard to answer because there are several perspectives. Note that CDI exists outside of JEE so this event makes sense even there. However, being able to fire this event yourself, you can use that to indicate whenever you think the container is ready (i.e. when other non-CDI services have also started?). Having it fired directly from Weld means we can only indicate when the CDI container started which isn't really that useful.

To me, the `Startup` event is ideally an entry point into the earliest time in the container lifecycle when you are safe to play around with all of its features. And this point can be very different in SE (where a bootstrap of CDI container is a job done) and in EE (where you also need to start other services and possibly only do that after CDI bootstraps).
However, this is not exactly what the javadoc states, it is purposefully very vague - https://github.com/eclipse-ee4j/cdi/blob/master/api/src/main/java/jakarta/enterprise/event/Startup.java